### DISCUSSION TITLE
How to explicitly mark a struct device_copiable

### DISCUSSION THREAD

**Discussion by Luigi-Crisci** [intel/llvm#8734] [Q&A] [answered]
Hello,
I have a struct that I use in a SYCL kernel targeting NVIDIA GPUs, but the compiler complains that it is not device_copiable. However, I know it is device_copiable as this kernel is a CUDA porting, which was working fine.
The SYCL doc specifies that
> In addition, the implementation may allow the application to explicitly declare certain class types as device copyable. If the implementation has this support, it must predefine the preprocessor macro SYCL_DEVICE_COPYABLE to 1,

In DPC++ this define exists, but I'm not able to extend it without having the compiler complain.
Which is the correct way of doing that?

**Comment by AlexeySachkov** [comment, marked as answer]
Hi @Luigi-Crisci,

You can find an example in our tests: [basic_tests/is_device_copyable.cpp](https://github.com/intel/llvm/blob/sycl/sycl/test/basic_tests/is_device_copyable.cpp).

Let's say you have your `struct`:
https://github.com/intel/llvm/blob/10e07a38ae57e2312a44cc7aa1a6999b2813fb14/sycl/test/basic_tests/is_device_copyable.cpp#L21-L26

Specializing `is_device_copyable` trait for it would look like:
https://github.com/intel/llvm/blob/10e07a38ae57e2312a44cc7aa1a6999b2813fb14/sycl/test/basic_tests/is_device_copyable.cpp#L63

You can also find examples in 3rd-party repos, such as [kokkos/kokkos](https://github.com/kokkos/kokkos/blob/aa1f48f3172069f212ed2c97b74b786999b26f17/core/src/SYCL/Kokkos_SYCL_Instance.hpp#L322), for example.
